extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ClickDelayFilter, makeClickDelaySpec} from './filters/click-delay';
+import {FilterType} from './filters/filter';
+import {assertConfig, TransportMode} from './config';
+import {isExperimentOn} from '../../../src/experiments';
+import {isJsonScriptTag, openWindowDialog} from '../../../src/dom';
+import {urlReplacementsForDoc} from '../../../src/services';
+import {user} from '../../../src/log';
+
+const TAG = 'amp-ad-exit';
+
+export class AmpAdExit extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {!./config.AmpAdExitConfig} */
+    this.config_ = {targets: {}, filters: {}};
+
+    this.defaultFilterSpecs_ = {
+      'minDelay': makeClickDelaySpec(1000),
+    };
+
+    this.clickDelayFilter_ = new ClickDelayFilter();
+
+    this.registerAction('exit', this.exit.bind(this));
+  }
+
+  /**
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   */
+  exit({args, event}) {
+    if (!isExperimentOn(this.win, 'amp-ad-exit')) {
+      user().warn('amp-ad-exit experiment is off.');
+      return;
+    }
+
+    event.preventDefault();
+    const targetName = args.target;
+    const target = this.config_.targets[targetName];
+    if (!target) {
+      user().error(TAG, `Exit target not found: '${targetName}'`);
+      return;
+    }
+    if (!this.filter_(
+            Object.keys(this.defaultFilterSpecs_), this.defaultFilterSpecs_,
+            event) ||
+        !this.filter_(target.filters, this.config_.filters, event)) {
+      return;
+    }
+    const substituteVariables =
+        this.getUrlVariableRewriter_(args, event, target);
+    if (target.trackingUrls) {
+      target.trackingUrls.map(substituteVariables)
+          .forEach(url => this.pingTrackingUrl_(url));
+    }
+    openWindowDialog(this.win, substituteVariables(target.finalUrl), '_blank');
+  }
+
+
+  /**
+   * @return {function(string): string}
+   */
+  getUrlVariableRewriter_(args, event, target) {
+    const vars = {
+      'CLICK_X': () => event.clientX,
+      'CLICK_Y': () => event.clientY,
+    };
+    const whitelist = {
+      'RANDOM': true,
+      'CLICK_X': true,
+      'CLICK_Y': true,
+    };
+    if (target.vars) {
+      for (const customVar in target.vars) {
+        if (customVar.startsWith('_')) {
```

Comment:
Use the [`startsWith` helper](https://github.com/ampproject/amphtml/blob/12ee5f23fb075c19d8544459ec838687573d581e/src/string.js#L54-L65), the method isn't available everywhere.